controller/forwarding/hyperlane_test.go

```diff
@@ -0,0 +1,475 @@
+			hypAttr: &forwardingtypes.HypAttributes{
+				TokenId:           usdnID,
+				DestinationDomain: 0,
+				Recipient:         make([]byte, 32),
+				CustomHookId:      make([]byte, 32),
+				GasLimit:          math.NewInt(1),
+				MaxFee:            sdk.NewInt64Coin("usdn", 1),
+			},
+			expError: "invalid transfer attributes",
+		},
+		{
+			name:         "error - when the token does not exist",
+			setup:        func(m *mocks.HyperlaneHandler) {},
+			transferAttr: transferAttr,
+			hypAttr: &forwardingtypes.HypAttributes{
+				TokenId:           usdnID,
+				DestinationDomain: 0,
+				Recipient:         make([]byte, 32),
+				CustomHookId:      make([]byte, 32),
+				GasLimit:          math.NewInt(1),
+				MaxFee:            sdk.NewInt64Coin("usdn", 1),
+			},
+			expError: "invalid Hyperlane forwarding",
+		},
+	}
+
+	for _, tC := range testCases {
+		t.Run(tC.name, func(t *testing.T) {
+			handler := mocks.HyperlaneHandler{Tokens: make(map[string]warptypes.WrappedHypToken)}
+
+			if tC.setup != nil {
+				tC.setup(&handler)
+			}
+
+			controller, err := forwarding.NewHyperlaneController(
+				log.NewNopLogger(),
+				handler,
+			)
+			require.NoError(t, err)
+
+			err = controller.ValidateForwarding(
+				context.Background(),
+				tC.transferAttr,
+				tC.hypAttr,
+			)
+
+			if tC.expError != "" {
+				require.ErrorContains(t, err, tC.expError)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestHandlePacket_Hyperlane(t *testing.T) {
+	usdnID := make([]byte, 32)
+	copy(usdnID, "usdn id")
+
+	testCases := []struct {
+		name     string
+		setup    func(*mocks.HyperlaneHandler) context.Context
+		packet   func() *types.ForwardingPacket
+		expError string
+	}{
+		{
+			name:     "error - when the attributes are valid",
+			packet:   func() *types.ForwardingPacket { return nil },
+			expError: "Hyperlane controller received nil packet",
+		},
+		{
+			name: "error - when forwarding is default values",
+			packet: func() *types.ForwardingPacket {
+				a, err := types.NewTransferAttributes(
+					core.PROTOCOL_IBC,
+					"channel-0",
+					"usdn",
+					math.NewInt(1),
+				)
+				require.NoError(t, err)
+
+				return &types.ForwardingPacket{
+					TransferAttributes: a,
+					Forwarding:         &core.Forwarding{},
+				}
+			},
+			expError: "error extracting Hyperlane forwarding",
+		},
+		{
+			name: "error - when transfer attributes are default values",
+			packet: func() *types.ForwardingPacket {
+				f, err := forwardingtypes.NewHyperlaneForwarding(
+					usdnID,
+					0,
+					make([]byte, 32),
+					make([]byte, 32),
+					"",
+					math.ZeroInt(),
+					sdk.NewCoin("usdn", math.ZeroInt()),
+					[]byte{},
+				)
+				require.NoError(t, err)
+
+				return &types.ForwardingPacket{
+					TransferAttributes: &types.TransferAttributes{},
+					Forwarding:         f,
+				}
+			},
+			expError: "error validating Hyperlane forwarding",
+		},
+		{
+			name: "error - when the remote transfer fails",
```

Comment:
nit: with this big amount of fields set it would be nice to add to the description why the transfer fails -- is it because of the zero value transferred?

Reply:
ohh it's just because of the context value `FailingContextKey`